### PR TITLE
update vdc collection to use a reference to the storage manager

### PR DESCRIPTION
**Pull request by Ryanmtate**
merge #31 and rebase on main before merging.

### CODE REVIEW

**Comment by Ryanmtate** [comment]
TODO: merge in @rschulman `async-storage-manager` branch and ensure vdc collection methods are async.

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
+.DS_Store
```
Comment (wyc) [inline .gitignore:1]: thank goodness

**Comment by Ryanmtate** [comment]
> TODO: merge in @rschulman `async-storage-manager` branch and ensure vdc collection methods are async.

We'll open a new PR with the async changes. There are additional changes that need to get populated through the wallet with some refactoring to be done upstream. This branch provides enough shared functionality to incorporate future changes later, versus in this PR.

File: src/vdc_collection.rs [hunk 2]
```diff
@@ -1,197 +1,193 @@
...
+    pub fn all_entries(
+        &self,
+        storage: &Arc<dyn StorageManagerInterface>,
+    ) -> Result<Vec<Key>, VdcCollectionError> {
+        Ok(storage
+            .list()?
+            .iter()
+            .filter(|key| key.0.contains(KEY_PREFIX))
+            .map(ToOwned::to_owned)
+            .collect())
```
Comment (rschulman) [inline src/vdc_collection.rs:162]: This replacement is no longer stripping `KEY_PREFIX` from the key. Is that desired? Are there other places that are relying on the prefix being stripped?

**Review by rschulman — CHANGES_REQUESTED** [review]
This looks broadly in good shape to me. Made a couple particular comments.

File: src/vdc_collection.rs [hunk 3]
```diff
@@ -1,197 +1,193 @@
...
-    fn id_to_key(&self, id: Uuid) -> Key {
-        self.str_to_key(&id.to_string())
-    }
-
-    /// Convert a string ref to a storage key.
-    fn str_to_key(&self, id: &str) -> Key {
-        Key(format!("{}{}", KEY_PREFIX, id))
+    /// Return the storage prefix, given a credential type.
+    pub fn storage_prefix(ctype: &CredentialType) -> String {
+        format!("{KEY_PREFIX}{}.", ctype)
```
Comment (cobward) [inline src/vdc_collection.rs:181]: I don't think we should add metadata about the credential to the key, it makes it more difficult to discover credentials, maintain and migrate at a later date.

File: src/vdc_collection.rs [hunk 4]
```diff
@@ -1,197 +1,193 @@
...
 }
 
 /// Verifiable Digital Credential Collection
 ///
 /// This is the main interface to credentials.
-pub struct VdcCollection {
-    storage: Box<dyn StorageManagerInterface>,
-}
+#[derive(Debug)]
+pub struct VdcCollection;
```
Comment (cobward) [inline src/vdc_collection.rs:74]: What's the motivation for this change?

Comment (rschulman) [reply src/vdc_collection.rs:74]: I'm not Ryan, but I think the reasoning here was to avoid dealing with the memory management of passing around the Box. Since VdcCollection is used from the Wallet, which stores its own `StorageManagerInterface`, we can pass in a reference to the storage manager when we need it instead of storing it permanently in the VdcCollection. See the changes to `add`, `get`, etc below.

File: src/vdc_collection.rs (hunk 3 above)
Comment (rschulman) [reply src/vdc_collection.rs:181]: The reasoning here is to be able to ask the VdcCollection to list all credentials of a given type. If we store the type of the credential in the storage path, we can filter on that later, in `all_entries_by_type`.

Comment (cobward) [reply src/vdc_collection.rs:181]: There's lots of things that we might need to filter on. If we need an effective way for lookup then we should maintain an external metadata map in my opinion.

File: src/vdc_collection.rs (hunk 4 above)
Comment (cobward) [reply src/vdc_collection.rs:74]: It's not clear to me how this is a problem, it just moves the reference to the storage manager from the constructor to the methods. AFAIK there's no complexity with memory management - when the VdcCollection is constructed we have a reference to the storage manager, and when it goes out of scope / is GC'd, the reference is destroyed. Is there some nuance I'm missing?

File: src/vdc_collection.rs (hunk 3 above)
Comment (rschulman) [reply src/vdc_collection.rs:181]: I mean, if I'm honest I think its unlikely that any individual device is going to store so many credentials that we can't just return an entire list and let the consumer deal with filtering it, but I'm not tied to this usage and I haven't had a reason to use it yet.

@Ryanmtate was this crucial for anything in particular that you want to speak up for it?

File: src/vdc_collection.rs (hunk 4 above)
Comment (rschulman) [reply src/vdc_collection.rs:74]: I'll let @Ryanmtate respond, but it also occurred to me that this might be a change related to moving away from using the callback interface from Uniffi, which required passing everything around in `Box`es, and instead being able to use regular borrows, but that storing a borrow in the struct itself would have caused an explosion of lifetimes. Again, I don't know for sure, though.

File: src/common.rs [hunk 5]
```diff
@@ -12,7 +12,7 @@ impl UniffiCustomTypeConverter for CredentialType {
         Ok(CredentialType::from(credential_type.as_str()))
     }
     fn from_custom(credential_type: Self) -> Self::Builtin {
-        credential_type.into()
+        (&credential_type).into()
```
[the remark this answers is not in the document]
Comment (Ryanmtate) [reply src/common.rs:15]: Yeah, I think I added `From<&CredentialType>` for String, which does not consume. But we can change to this so that it is consumed.
